proposal for docs site generation to use branch deploys (#3729)
projectcalico/calico: RELEASING.md

1. Create a new branch off of the latest master.

   ```
   git checkout -b release-vX.Y
   ```
   
1. Enable branch deployments for this branch on netlify by adding branch name to `Deploy contexts` in site settings.
   
1. Commit your changes. For example:

   ```
   git commit -m "build vX.Y candidate"
   ```
   
1. Open a pull request to the upstream release-vX.Y branch. Get it reviewed and merged. 

> 

Reviewer: There are no changes anymore since we no longer update netlify.toml above, so there's nothing to commit. Can this be removed?

Author: sadly..we need a commit here to trigger the build (since the candidate branch is just added to branch deploys). I'm planning to ask the user to change the `RELEASE_VERSION` in the `netlify.toml` instead

   ```
   git checkout -b release-candidate-vX.Y
   ```
   
1. Enable branch deployments for this branch on netlify by adding branch name to `Deploy contexts` in site settings.

1. Add a new stanza for the build of this branch in [netlify.toml](netlify.toml). 

    ```toml

>    [build.<branch-name>]

Reviewer: ```suggestion
   [build.release-vX.Y]
```

>       make bin/helm

Reviewer: Isn't this supposed to be in triple quotes under a 'build' field?

      export PATH=$PATH:$(pwd)/bin
      jekyll build --config _config.yml,$(pwd)/netlify/_config_noindex.yml --baseurl /vX.Y
      jekyll build --config _config.yml,$(pwd)/netlify/_config_noindex.yml,$(pwd)/netlify/_manifests_only.yml --baseurl /vX.Y --destination _site/vX.Y
    ```
   
1. Commit your changes. For example:

   ```
   git commit -m "build vX.Y candidate"
   ```
   
1. Open a pull request to the upstream release-candidate-vX.Y branch. Get it reviewed and merged. 

1. After ensuring that the branch deployment is finished, in production branch's [netlify.toml](netlify.toml), add a new stanza for proxy to branch deploy site of release candidate.

   ```toml
   [[redirects]]
     from = "/vX.Y/*"

>      to = "https://<branch-name>--calico-test.netlify.app/:splat"

Reviewer: lets use `release-vX.Y` instead of `<branch-name>`.

And lets use `calico` instead of `calico-test`.

```suggestion
     to = "https://release-vX.Y--calico.netlify.app/:splat"
```

   git add _data/release-notes/<VERSION>-release-notes.md
   ```

1. In [netlify.toml](netlify.toml):
      
      1. Set the `RELEASE_VERSION` environment variable to new version.
      

> 1. On netlify update `Production branch` in Deploy contexts to release-vX.Y in docs.projectcalico.org site settings.

Reviewer: does this need to be done before or after the surrounding changes are merged? If before, can you move it up to be step 1? And if after, can you move it down after step 10?

Having it sit in between is a bit odd because the user is in the middle of  building a PR to master

Author: I think that's doable. I added it in between because doing it before the commit will trigger the build. But we can ask the user to trigger the build manually after updating the production branch.

   git add _data/release-notes/<VERSION>-release-notes.md
   ```

1. In [netlify.toml](netlify.toml):
      

>       1. Set the `RELEASE_VERSION` environment variable to new version.

Reviewer 2: Do we still need this step since we already do this on the branch when the release branch is created?

   make release-publish
   ```

1. Merge the PR. This will cause docs.projectcalico.org to be updated (after a few minutes). Validate that everything looks correct.

> ## Adding the previous release to docs.projectcalico.org/archive

Reviewer: Do we need to do this before we promote the candidate release? I'm worried there will be a blip where the 'previous release' is not built or published because it's no longer the production branch and it hasn't been added as a 'previous release' branch.

I'm wondering what netlify will do if a branch is both listed as the production branch _and_ listed as a branch build. A quick check of the UI and it doesn't seem like they let you specify it in both.

How hard would it be to prevent this blip? If it's too hard, I think we just note that this is one small but unfortunate side effect of this new process

Author: yeah.. it doesn't let us specify a branch as both. I don't see an easy way out unfortunately.. We can  make the production branch commit, a little earlier while doing the release to shorten the time.. but it'll still be a branch build away in terms of time.

1. Enable branch deployments for previous release branch on netlify by adding branch name to `Deploy contexts` in site settings.
 

> 1. Checkout to previous release branch. Add a commit to the branch, an empty commit if necessary (to trigger the branch deploy). 

Reviewer: I really wish this wasn't required but in your testing you couldn't find another way to trigger the new branch build, right?

Author: yeah.. I couldn't.. there are [build hooks](https://docs.netlify.com/configure-builds/build-hooks/#parameters) but they only work for branches that are deployed at least once

1. Open a pull requests to upstream previous release branch, get it reviewed and merged. This triggers branch deployment for the previous release.

> 1. Ensure branch deployment is done by visiting to `<previous-release-branch-name>--calico.netlify.app`

Reviewer: Should we note that the CSS won't load here but that is OK?

Author: right, I'll add this.

1. Check out to production branch. 
   ```
   git checkout release-vX.Y
   ```

> 1. Add a new stanza for proxy for the previous release.

Reviewer: ```suggestion
1. Add a new stanza to [netlify.toml](netlify.toml) to configure the proxy to the previous release.
```

Author: fixed

1. Commit your changes. For example:
1. Ensure branch deployment is done by visiting to `<previous-release-branch-name>--calico.netlify.app`

1. Check out to production branch. 
   ```
   git commit -m "Promote vX.Y.Z to latest"
   git checkout release-vX.Y
   ```

1. Push your branch and open a pull request to the upstream master branch. Get it reviewed and wait for it to pass CI.

## Adding the previous release to archive.docs.projectcalico.org
1. Add a new stanza for proxy for the previous release.

1. Checkout latest master.

   ```
   git checkout master
   ```toml
    [[redirects]]
      from = "/archive/<previous-release>/*"
      to = "https://<previous-release-branch>--calico-test.netlify.app/:splat"
      status = 200
      force = true
      headers = {X-From = "Netlify"}
   ```

> 1. Add the previous release to the top of `_data/archive.yml`

Reviewer: I think we still need to add the previous release to the top of this file as it is used to generate the list of releases on the /releases page

   make release-publish
   ```

1. Merge the PR. This will cause docs.projectcalico.org to be updated (after a few minutes). Validate that everything looks correct.

> 1. Remove branch deploy for the candidate branch in docs.projectcalico.org site settings.

Reviewer: I thought we left the branch deploy in place for the branch as it transitions from candidate to latest?

## Adding the previous release to docs.projectcalico.org/archive

> 1. Enable branch deployments for previous release branch on netlify by adding branch name to `Deploy contexts` in site settings.

Reviewer: Yeah we wouldn't have to re-enable the branch builds if we left them on in the last section
